core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicHttpFunctionImpl.java

```diff
@@ -78,23 +88,67 @@
+                parameters[i] = message;
             }
         }
         return parameters;
     }
 
     @Override
     public Object serve(ServiceRequestContext ctx, HttpRequest req, Map<String, String> args) throws Exception {
-        Object[] parameters = parameters(ctx, req, args);
+        if (aggregationRequired) {
+            if (CompletionStage.class.isAssignableFrom(method.getReturnType())) {
+                return req.aggregate()
+                          .thenComposeAsync(msg -> (CompletionStage<?>) invokeMethod(ctx, req, args, msg),
+                                            ctx.contextAwareEventLoop());
+            } else if (isAsynchronous) {
+                return req.aggregate()
+                          .thenApplyAsync(msg -> invokeMethod(ctx, req, args, msg),
+                                          ctx.contextAwareEventLoop());
+            } else {
+                return req.aggregate()
+                          .thenApplyAsync(msg -> invokeMethod(ctx, req, args, msg),
+                                          ctx.blockingTaskExecutor());
+            }
+        }
+
         if (isAsynchronous) {
+            return invokeMethod0(ctx, req, args, null);
+        } else {
+            return CompletableFuture.supplyAsync(
+                    () -> invokeMethod0(ctx, req, args, null), ctx.blockingTaskExecutor());
+        }
+    }
+
+    private Object invokeMethod(ServiceRequestContext ctx, HttpRequest req,
+                                Map<String, String> args, AggregatedHttpMessage message) {
+        ServiceRequestContext wrappedCtx = wrap(ctx, message);
+        try (SafeCloseable ignored = RequestContext.push(wrappedCtx, false)) {
+            return invokeMethod0(wrappedCtx, req, args, message);
+        }
+    }
+
+    private Object invokeMethod0(ServiceRequestContext ctx, HttpRequest req, Map<String, String> args,
+                                 @Nullable AggregatedHttpMessage message) {
+        try {
+            Object[] parameters = parameters(ctx, req, args, message);
             return method.invoke(object, parameters);
+        } catch (IllegalAccessException | InvocationTargetException e) {
+            throw new IllegalStateException(e);
         }
-        return CompletableFuture.supplyAsync(() -> {
-            try {
-                return method.invoke(object, parameters);
-            } catch (IllegalAccessException | InvocationTargetException e) {
-                throw new IllegalStateException(e);
+    }
+
+    private static ServiceRequestContext wrap(ServiceRequestContext context,
+                                              AggregatedHttpMessage message) {
+        requireNonNull(context, "context");
+        requireNonNull(message, "message");
+        return new ServiceRequestContextWrapper(context) {
+            @Override
+            @SuppressWarnings("unchecked")
+            public <T> T request() {
+                return (T) message;
```

Comment:
You don't need to wrap because `request()` is expected to return the 'original request'.